### DISCUSSION TITLE
Navigation recalculation optimization

### DISCUSSION THREAD

**Discussion by 0709wiwiwi** [devemux86/cruiser#191] [Q&A] [answered]
In navigation, multiple recalculated return commands are annoying. 
If you ignore the first return command, an alternative is expected soon.

Principle.

"_**Don't send me back to the green start pin location because I just came from there**_"

**Comment by devemux86** [comment]
@0709wiwiwi thanks for the idea!

It is already in my list for implementation. :slightly_smiling_face:

**Comment by devemux86** [comment]
I reported that BRouter heading does not work in the BRouter repository:
- https://github.com/abrensch/brouter/issues/589

---

Kurviger heading also does not work, as their focus is on routing, not navigation:
- https://forum.kurviger.de/t/routing-heading/5289

---

It seems the only reliable solution is for Cruiser to implement routing heading on top of the routers.

**Comment by devemux86** [comment]
Implementation was easy and will be available in next Cruiser.

Solving for good the navigation rerouting direction within Cruiser.
(for BRouter and all routing services)

**Reply by devemux86** [reply, marked as answer]
> Solving for good the navigation rerouting direction within Cruiser.
(for BRouter and all routing services)

Implemented in [Cruiser 3.0.30](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Reply by 0709wiwiwi** [reply]
The auto rerouting (recalculation) after frequent cycling tests works perfectly.
Still this question remains: Which nogo diameter to prevent a return is placed.

**Reply by devemux86** [reply]
Right now it's a fixed radius of 20m, same distance back.

After suggestion by @afischerdev we use weighted no-go (no strict) to avoid failed routing:
- https://github.com/abrensch/brouter/issues/589#issuecomment-1660476991

**Reply by 0709wiwiwi** [reply]
Emux can you check please ?
The autorerouting by a cycle bike profile stops the multiple return routing results  after a recalculation; 
The autorerouting by a foot profile does not ?  (Both tested by the Brouter offline)

**Reply by devemux86** [reply]
The profiles work similarly in the rerouting.
Also after failed rerouting, the next one happens in 30".

> stops the multiple return routing results after a recalculation

What do you mean? After a rerouting why not use the new route?
Can you post a route example and the profile?

**Reply by devemux86** [reply]
One difference is that pedestrian profiles (walking, running, hiking) do not use direction in the rerouting.
i.e. rerouting is permitted to direct the pedestrian in the opposite direction.

Motor and bike profiles use the user's direction in rerouting.

**Reply by devemux86** [reply]
@0709wiwiwi thanks, you mention "cycle" but the profile is "Walking"?

I will enable the heading also in pedestrian profiles to avoid the backwards rerouting.

**Reply by 0709wiwiwi** [reply]
> Right now it's a fixed radius of 20m.

Does this mean you so have to add that 20m radius, this to know the real rigger distance ? 
With thus the shortest possible trigger distance (by foot profiles) is 20m+1m= 21 meters.

**Reply by devemux86** [reply]
Rerouting distance is different, it is used with the selected value.
No-go areas are not related with it.

**Reply by devemux86** [reply]
As min? What the min / default should be?

**Reply by 0709wiwiwi** [reply]
Default 20m(Foot)

**Comment by devemux86** [comment]
> I will enable the heading also in pedestrian profiles to avoid the backwards rerouting.

Implemented in [Cruiser 3.0.32](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

Also increased the pedestrian default off route distance to 20m (min 10m).
